### ISSUE TITLE
EPIC: Create Inventory of and then Fix Any Memory Leak and Memory Overrun Issues

### ISSUE THREAD

**Issue by john-boyer-phd** [graph-algorithms/edge-addition-planarity-suite#59]
- [x] Set up a memory checker tool such as `Leaks` on the Mac first (Windows and Linux may be done later, with a different toolset). (see #81 )
- [x] Investigate whether the memory checker tool can be meaningfully run with the release executable versus only with the debug version. (see [this comment](https://github.com/graph-algorithms/edge-addition-planarity-suite/issues/59#issuecomment-2190017861))
- [x] Make any necessary changes to make and/or configuration (`.ini` file generated using `configparser` module) files to allow rapid inclusion/exclusion of various memory leak and overrun detection features.
- [ ] Add a page to the wiki to describe how to set up and run the `Leaks` tool with planarity and how to do the inclusion/exclusion of the various features.
- [x] From the menu or the command-line, run the following
  - [x] all algorithms (P, D, O, 2, 3, 4) using specific graph mode, 
  - [x] random graphs with one small graph, 
  - [x] the random max planar graph generator,
  - [x] the random nonplanar graph (maximal planar + edge) generator,
  - [x] the graph converter, and 
  - [x] test all graphs 
  to see if there are any memory leaks or memory overruns on a file containing a single small graph (e.g., a 6-vertex or an 8-vertex graph). For each memory issue, add a story item to this list right below this item to fix the issue.
- [x] Generate all 8-vertex graphs into a single `graph6` file. This size has been chosen to ensure we trigger the memory expansion code that `gp_Read()` invokes (via `gp_DynamicAddEdge()`; see [this comment](https://github.com/graph-algorithms/edge-addition-planarity-suite/issues/64#issuecomment-2089249300)). 
- [x] For the given platform (Mac with Leaks), run the memory checker tool on the `planarity` test all graphs function for each of the six algorithms (pdo234) on the set of 8-vertex graphs. For any memory leak or memory overrun issue(s), add issue item(s) directly below this item to fix the issue(s).
- [x] Do some calculations to figure out how long it would take to run all 10-vertex graphs through the memory leak and overrun testing, then decide how many 10-vertex graphs to do and add an issue.
    * Running tests for all graphs of order 10 took a short amount of time; only `TestAllGraphs` for `K_{3, 3}` search was the outlier at ~38 minutes: [2024-07-09_09-33-40_n10.mALL.g6_K10.g6.zip](https://github.com/user-attachments/files/16147549/2024-07-09_09-33-40_n10.mALL.g6_K10.g6.zip)
    * #82 

Once all memory issues have been resolved on the Mac with Leaks, then we can consider trying Linux and Windows, which may be done as separate epics.

**Comment by wbkboyer** [comment]
There's a bit of a bummer; it seems from [Valgrind Supported Platforms](https://valgrind.org/info/platforms.html) that there is no Windows support for Valgrind :( One response to a StackOverflow post [gives examples of OSS and free tools](https://stackoverflow.com/a/6580332/5456879) for memory analysis of Windows applications.

<details>
<summary>Investigation: can you check a Release build for memory issues?</summary>

In order to use Valgrind on Linux or Leaks on MacOS, you have to compile the project using the right compiler and linker flags to get debug info. [This comment](https://github.com/graph-algorithms/edge-addition-planarity-suite/issues/54#issuecomment-2061885027) on #54 is the result of my investigation of different compile options for Debug and Release builds; to get debug info when using different compiler toolchains:
* For `gcc`, [Debugging Options](https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html) indicates that `-g_level_` (e.g. `-g3`) requests debugging information at the specified level
* For `clang`, [Controlling Debug Information](https://clang.llvm.org/docs/UsersManual.html#id61) indicates the [`-g`](https://clang.llvm.org/docs/UsersManual.html#cmdoption-g) option generates complete debug info, and [`-fdebug-macro`](https://clang.llvm.org/docs/UsersManual.html#cmdoption-fdebug-macro) generates debug info for preprocessor macros

[This StackOverflow response](https://stackoverflow.com/a/24701207/5456879) describes how without the `-g` compiler option, the report produced by `valgrind` (and likely similarly for `leaks`) won't contain useful information (e.g. line number on which leak occurs).

Another important note is that using the compiler option to optimize code at level 2 (possibly even level 1) may cause false-positives and false-negatives when doing memcheck (see [valgrind](https://valgrind.org/docs/manual/manual-core.html) for full details).

What flags are being used for the current "Release" builds?
* The VSCode Release build configurations for `clang` (MacOS) and `gcc` (MinGW/Linux) in `tasks.json` have `-g` and `-O3` flags, and don't include `-DDEBUG` nor `-Wall`
* When we [Make the distribution](https://github.com/graph-algorithms/edge-addition-planarity-suite?tab=readme-ov-file#making-the-distribution) for the `edge-addition-planarity-suite`, [this StackOverflow post](https://stackoverflow.com/a/16672757/5456879) says that `autoconf` (run as a step in `autogen.sh`) will use the following default flags:
  > If using the GNU C compiler, set shell variable GCC to ‘yes’. If output variable CFLAGS was not already set, set it to -g -O2 for the GNU C compiler (-O2 on systems where GCC does not accept -g), or -g for other compilers

_**N.B.**_ Currently, gnu.org is down, so I can't read the current docs firsthand

These two pieces of information indicate that in order to find memory leaks using Valgrind, you need to compile with `-g` and *at most* `-O` for code optimization to reduce the chance of false-negatives and false-positives. Additionally, running with `-Wall` may catch issues at a higher level of abstraction that might be missed if we use the higher optimization levels.

</details>

TL;DR Yes, but in order to get any useful information from running these tools on a Release build, you need to have compiled with `-g` flag at the very least, and should use `-O` (Rather than default `-O2` coming from `autoconf`, or `-O3` in VSCode `tasks.json`) to reduce the chance of false-positives and false-negatives.

**Comment by wbkboyer** [comment]
Before using `leaks` on MacOS, it is recommended that one first
```
export MallocStackLogging=1
```
so that we have access to the function call stack information for when memory is allocated.

<details>
<summary>Comparing output with and without MallocStackLogging=1</summary>

Without `MallocStackLogging=1`:
```
...
leaks Report Version: 4.0
Process 20815: 206 nodes malloced for 17 KB
Process 20815: 13 leaks for 2640 total leaked bytes.

    13 (2.58K) ROOT LEAK: 0x12a604f80 [352]
       1 (800 bytes) 0x12a605340 [800]
       2 (416 bytes) 0x60000301c0c0 [16]
          1 (400 bytes) 0x12a605660 [400]
       1 (336 bytes) 0x12a6051f0 [336]
       1 (272 bytes) 0x12a6050e0 [272]
       1 (144 bytes) 0x600000018120 [144]
       2 (128 bytes) 0x60000301c0d0 [16]
          1 (112 bytes) 0x600001918070 [112]
       2 (96 bytes) 0x60000301c0a0 [16]
          1 (80 bytes) 0x6000011180f0 [80]
       2 (96 bytes) 0x60000301c0b0 [16]
          1 (80 bytes) 0x600001118140 [80]
...
```

With `MallocStackLogging=1`:
```
...
leaks Report Version: 4.0, multi-line stacks
Process 20842: 206 nodes malloced for 17 KB
Process 20842: 13 leaks for 2640 total leaked bytes.

STACK OF 1 INSTANCE OF 'ROOT LEAK: <malloc in gp_New>':
5   dyld                                  0x183f0e0e0 start + 2360
4   planarity                             0x104212eac commandLine + 724  planarityCommandLine.c:0
3   planarity                             0x1042158ec TestGraphFunctionality + 560  planarityTestGraphFunctionality.c:149
2   planarity                             0x104215bac testAllGraphs + 200  planarityTestGraphFunctionality.c:275
1   planarity                             0x10420e0bc gp_New + 20  graphUtils.c:92
0   libsystem_malloc.dylib                0x1840d0adc _malloc_zone_malloc_instrumented_or_legacy + 264 
====
    13 (2.58K) ROOT LEAK: <malloc in gp_New 0x159e04f80> [352]
       1 (800 bytes) <calloc in _InitGraph 0x159e05340> [800]
       2 (416 bytes) <malloc in sp_New 0x6000022a00c0> [16]
          1 (400 bytes) <malloc in sp_New 0x159e05660> [400]
       1 (336 bytes) <calloc in _InitGraph 0x159e051f0> [336]
       1 (272 bytes) <calloc in _InitGraph 0x159e050e0> [272]
       1 (144 bytes) <calloc in _InitGraph 0x6000012a4120> [144]
       2 (128 bytes) <malloc in sp_New 0x6000022a00d0> [16]
          1 (112 bytes) <malloc in sp_New 0x600000ba4070> [112]
       2 (96 bytes) <malloc in LCNew 0x6000022a00a0> [16]
          1 (80 bytes) <malloc in LCNew 0x6000003a40f0> [80]
       2 (96 bytes) <malloc in LCNew 0x6000022a00b0> [16]
          1 (80 bytes) <malloc in LCNew 0x6000003a4140> [80]
...
```

</details>

The [Apple Developer docs on Leaks](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/ManagingMemory/Articles/FindingLeaks.html#//apple_ref/doc/uid/20001883-100980) link to recommendations for [Enabling Malloc Debugging Features](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/ManagingMemory/Articles/MallocDebug.html#//apple_ref/doc/uid/20001884-CJBJFIDD), which indicate the following additional environment variables might be useful to enable:

Variable | Description
-- | --
MallocStackLoggingNoCompact | This option is similar to MallocStackLogging but makes sure that all allocations are logged, no matter how small or how short lived the buffer may be.
MallocScribble | If set, free sets each byte of every released block to the value 0x55.
MallocPreScribble | If set, malloc sets each byte of a newly allocated block to the value 0xAA. This increases the likelihood that a program making assumptions about freshly allocated memory fails.
MallocGuardEdges | If set, malloc adds guard pages before and after large allocations.
MallocCheckHeapStart | Set this variable to the number of allocations before malloc will begin validating the heap. If not set, malloc does not validate the heap.
MallocCheckHeapEach | Set this variable to the number of allocations before malloc should validate the heap. If not set, malloc does not validate the heap.

However, doing so will significantly slow down execution of `leaks`.

**Comment by wbkboyer** [comment]
_**TL;DR**_ You either get incomplete logs from `leaks`, or the process hangs :( 

## Troubleshooting `leaks` output not being completely captured
 
When `leaks` is run using `subprocess.run()` with `stdout` and `stderr` routed to separate output files *misses* some of the log messages. For example, I tested by adding a double-free to `planarity` and recompiling before running:
```
MallocStackLogging=1 MallocStackLoggingNoCompact=1 MallocScribble=1 MallocPreScribble=1 MallocGuardEdges=1 MallocCheckHeapStart=1 MallocCheckHeapEach=1 MallocErrorAbort=1 MallocCheckHeapAbort=1 leaks -atExit -- ../../../Debug/planarity -s -p n6.m13.1.g6 n6.m13.1.s.p.out.txt
```

And among many other log messages, I witnessed:
```
        planarity(87264,0x1f049cc00) malloc: *** error for object 0x158608000: pointer being freed was not allocated
        planarity(87264,0x1f049cc00) malloc: *** set a breakpoint in malloc_error_break to debug
```

However, when I run leaks from a `python` script using `subprocess.run()`:
```
     with open(
            stdout_outfile_path, "w", encoding="utf-8"
        ) as stdout_outfile, open(
            stderr_outfile_path, "w", encoding="utf-8"
        ) as stderr_outfile:
            # running leaks processes hang if I set stderr=subprocess.PIPE, so
            # instead of routing stderr to subprocess.STDOUT, might as well
            # capture stderr in a separate file.
            full_args = ["leaks", "-atExit", "--"] + command_args
            subprocess.run(
                full_args,
                stdout=stdout_outfile,
                stderr=stderr_outfile,
                env=leaks_env,
                check=False,
                cwd=cwd,
            )
```

Only *some* of the logs from `leaks` are present in the `stdout_outfile`.

### Various ways of calling `leaks` directly

I tried to see if I could get the data output to `stdout` and `stderr` by `leaks`:
1. Calling `subprocess.run()` with `capture_output=True` and *not* setting `stdout` and `stderr` redirection causes the process to hang
2. Calling `subprocess.run()` with `stdout=subprocess.PIPE` and `stderr=subprocess.STDOUT` causes the process to hang, so we never get the `CompletedProcess` instance to access its attributes `stdout` and `stderr`
3. Creating a `subprocess.Popen` instance and redirecting stdout and stderr to `subprocess.PIPE`, then trying to capture the output of having run the process with `out, err = process.communicate()` hangs.
4. Creating a `subprocess.Popen()` instance with `stdout` and `stderr` redirected to `subprocess.PIPE` and then directly trying to `process.stdout.read()` and `process.stdout.read()`

I also verified that the output contents should not exceed the system IO buffer size (as warned in the `subprocess` documentation [here](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr)), so that shouldn't be causing a deadlock.

### Calling `leaks` via `stdbuf` to try to control stream handling

I wanted to try to use [`stdbuf`](https://man.cx/stdbuf) to change how `leaks` was handling output buffering:
> If MODE is ’L’ the corresponding stream will be line buffered. This option is invalid with standard input.
> If MODE is ’0’ the corresponding stream will be unbuffered.

I ran:
1. `"stdbuf -o0 -e0 leaks -atExit --" + " ".join(command_args)` with:
    - `subprocess.run()` with `capture_output=True` and `bufsize=1` first with `text=True` and then with `universal_newlines=True`
    - `subprocess.Popen()` instance with `stdout` and `stderr` redirected to `subprocess.PIPE` and using `bufsize=1` first with `text=True` and then with `universal_newlines=True`
 I even added `sys.stdout.flush()` and `sys.stderr.flush()` before and after these calls, but still it hung.
2. `"stdbuf -oL -eL leaks -atExit --" + " ".join(command_args)` with:
    - `subprocess.run()` with `capture_output=True`
    - `subprocess.Popen()` instance with `stdout` and `stderr` redirected to `subprocess.PIPE` and then trying to capture the output of having run the process with `out, err = process.communicate()`
    - `subprocess.Popen()` instance with `stdout` and `stderr` redirected to `subprocess.PIPE` and then directly trying to `process.stdout.read()` and `process.stdout.read()`


---

As an aside, I found a possible new tool to help check for memory issues at compile time: [LeakSanitizer](https://clang.llvm.org/docs/LeakSanitizer.html). However, I think we'd need to install a separate clang compiler (i.e. `LeakSanitizer` isn't supported by the `clang` bundled with Xcode tools)

**Comment by wbkboyer** [comment]
_**TL;DR**_ Memory overruns don't seem to be detected by `leaks`???

At the same time that I was investigating the above issue, I added some code to `planarity` in `_InitGraph()` and `gp_Free()` that would allocate memory for two new dummy attributes added to the `baseGraphStructure` `struct`, each being a `char *` pointing to `2*1048576` bytes of memory. Then I proceeded to write to the address after the end of memory allocated for the first attribute (which I confirmed via `printf()` statments was changing the character at the address of (i.e. the first index of) the second attribute). As well, I wrote 100000 characters after the end of the memory allocated for the second attribute. Finally, I included the double-free.

I set the following environment variables for running `leaks` from the command line:
```
MallocStackLogging=1
MallocStackLoggingNoCompact=1
MallocScribble=1
MallocPreScribble=1
MallocGuardEdges=1
MallocCheckHeapStart=1
MallocCheckHeapEach=1
MallocErrorAbort=1
MallocCheckHeapAbort=1
```

I saw *no logs* indicating that memory had been overrun, only that there was a double-free:
```
planarity(6934,0x1f049cc00) malloc: *** error for object 0x128608000: pointer being freed was not allocated
planarity(6934,0x1f049cc00) malloc: *** set a breakpoint in malloc_error_break to debug
```

[overrun_not_detected.log](https://github.com/user-attachments/files/16114579/overrun_not_detected.log)

**Comment by john-boyer-phd** [comment]
> _**TL;DR**_ Memory overruns don't seem to be detected by `leaks`???
> 
> At the same time that I was investigating the above issue, I added some code to `planarity` in `_InitGraph()` and `gp_Free()` that would allocate memory for two new dummy attributes added to the `baseGraphStructure` `struct`, each being a `char *` pointing to `2*1048576` bytes of memory. Then I proceeded to write to the address after the end of memory allocated for the first attribute (which I confirmed via `printf()` statments was changing the character at the address of (i.e. the first index of) the second attribute). As well, I wrote 100000 characters after the end of the memory allocated for the second attribute. Finally, I included the double-free.
> 
> I set the following environment variables for running `leaks` from the command line:
> 
> ```
> MallocStackLogging=1
> MallocStackLoggingNoCompact=1
> MallocScribble=1
> MallocPreScribble=1
> MallocGuardEdges=1
> MallocCheckHeapStart=1
> MallocCheckHeapEach=1
> MallocErrorAbort=1
> MallocCheckHeapAbort=1
> ```
> 
> I saw _no logs_ indicating that memory had been overrun, only that there was a double-free:
> 
> ```
> planarity(6934,0x1f049cc00) malloc: *** error for object 0x128608000: pointer being freed was not allocated
> planarity(6934,0x1f049cc00) malloc: *** set a breakpoint in malloc_error_break to debug
> ```
> 
> [overrun_not_detected.log](https://github.com/user-attachments/files/16114579/overrun_not_detected.log)

Wow, clearly guard edges doesn't mean what we think it should mean :-(

**Comment by wbkboyer** [comment]
The problem is worse than first surmised: there seems to be some issue with redirecting all of the output of `leaks` to file. If you allow it to spew everything out to the terminal, all logs are present.

<details>
<summary>Investigation details</summary>

If `stdout` and `stderr` are allowed to default to `None` for `subprocess.run()`:
```
            full_args = ["leaks", "-atExit", "--"] + command_args
            subprocess.run(
                full_args,
                env=leaks_env,
                check=False,
                cwd=cwd,
            )
```

Then all the processes spew out the results of running `planarity_leaks_orchestrator.py` to the terminal window, 
```
% python3 planarity_leaks_orchestrator.py 
```

I see **all** the logs, including those indicating the double free in the terminal window (albeit interleaved due to `multithreading`).
[planarity_leaks_orchestrator-noRedirect.txt](https://github.com/user-attachments/files/16132682/planarity_leaks_orchestrator-noRedirect.txt)

If I use `>` to redirect the output from running
```
python3 planarity_leaks_orchestrator.py > planarity_leaks_orchestrator-redirectedLogs.txt
```
Only **some** of the logs appear in `foo.txt` (which don't appear in the terminal window), where a whole whack of other logs appear only in the terminal window.
[planarity_leaks_orchestrator-redirectedLogs.txt](https://github.com/user-attachments/files/16132689/planarity_leaks_orchestrator-redirectedLogs.txt)
[planarity_leaks_orchestrator-redirectToFile.txt](https://github.com/user-attachments/files/16132690/planarity_leaks_orchestrator-redirectToFile.txt)

---

This still happens if I run `leaks` directly and redirect to the output file (even using `stdbuf -o0 -e0` to ensure the streams would be unbuffered):
```
% MallocStackLogging=1 MallocStackLoggingNoCompact=1 MallocScribble=1 MallocPreScribble=1 MallocGuardEdges=1 MallocCheckHeapStart=1 MallocCheckHeapEach=1 MallocErrorAbort=1 MallocCheckHeapAbort=1 stdbuf -o0 -e0 leaks -atExit -- ../../../Debug/planarity -s -p n6.m13.1.g6 n6.m13.1.s.p.out.txt > leaks_direct-redirectedLogs.txt
```

The logs like
```
planarity(7461,0x202ba0c00) malloc: *** error for object 0x120608000: pointer being freed was not allocated
planarity(7461,0x202ba0c00) malloc: *** set a breakpoint in malloc_error_break to debug
```
Are shown in the terminal window, but are not captured in `leaks_direct-redirectedLogs.txt`.
[leaks_direct-redirectedLogs.txt](https://github.com/user-attachments/files/16132870/leaks_direct-redirectedLogs.txt)
[leaks_direct-redirectToFile.txt](https://github.com/user-attachments/files/16132871/leaks_direct-redirectToFile.txt)

---

I also tried to see if this was an issue with `zsh`, the default shell, by changing the default shell on MacOS to `bash`:
```
chsh -s $(which bash)
```
then completely exiting Terminal and re-opening and trying again, but the same issue manifested.

N.B. I switched back to `zsh` using
```
chsh -s $(which zsh)
```

---

I even tried to redirect all file descriptors to separate files:
```
% MallocStackLogging=1 MallocStackLoggingNoCompact=1 MallocScribble=1 MallocPreScribble=1 MallocGuardEdges=1 MallocCheckHeapStart=1 MallocCheckHeapEach=1 MallocErrorAbort=1 MallocCheckHeapAbort=1 stdbuf -o0 leaks -atExit -- ../../../Debug/planarity -s -p n6.m13.1.g6 n6.m13.1.s.p.out.txt 0 >foo0.txt 1>foo1.txt 2>foo2.txt 3>foo3.txt 4>foo4.txt 5>foo5.txt 6>foo6.txt 7>foo7.txt 8>foo8.txt 9>foo9.txt
```
But I was still missing logs in the redirected output files (`foo0.txt` and `foo1.txt` contained the same lines from `stdout`, `foo2.txt` contained `stderr` output, and all other files were empty)

</details>

**Comment by wbkboyer** [comment]
@m-hemmings introduced me to the [`script`](https://man.freebsd.org/cgi/man.cgi?query=script&apropos=0&sektion=1&manpath=macOS+14.3.1&arch=default&format=html) command, which I was able to verify successfully captured *ALL* logs from `leaks` running `planarity`. I implemented a workaround that works with `multiprocessing` and `subprocess`: if I wrap the call to `leaks` using the `script` command:

```
        outfile_path = leaks_outfile_basename.with_suffix(
            leaks_outfile_basename.suffix + ".log"
        )
        if cwd:
            outfile_path = Path.joinpath(cwd, outfile_path)

        full_args = [
            "script",
            f"{outfile_path}",
            "leaks",
            "-atExit",
            "--",
        ] + command_args
        subprocess.run(
            full_args,
            # Must default to None for stdout and stdin so that script utility
            # will capture *all* output; see #59 for investigation
            env=leaks_env,
            check=False,
            cwd=cwd,
        )
```
I capture *ALL LOGS* from `leaks` running `planarity` with various parameterizations!

However, one must run the `planarity_leaks_orchestrator.py` as follows:
```
python3 planarity_leaks_orchestrator.py >/dev/null 2>&1
```

Otherwise you'll also get all the output from the various threads interleaved in the terminal window as well.
